src/lib/set_operations.js

```diff
@@ -0,0 +1,35 @@
+/**
+* Copyright 2012-2018, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+
+/*
+ * Computes the set difference of two arrays.
+ *
+ * @param {array} a
+ * @param {array} b
+ * @returns all elements of a that are not in b.
+ *      If a is not an array, an empty array is returned.
+ *      If b is not an array, a is returned.
+ */
+function difference(a, b) {
+    if(!Array.isArray(a)) {
+        return [];
+    }
+    if(!Array.isArray(b)) {
+        return a;
+    }
+    return a.filter(function(e) {
+        return b.indexOf(e) < 0;
```

Comment:
Looks like `Lib.difference` is only called on array of number. We could take advantage of that by using a hash table instead of `indexOf` which can be slow for arrays with length > 1e5.

Something like:

```js
var small;
var large;
var hash = {};
var out = [];
var i;

if(a.length < b.length) {
  small = a;
  large = b;
} else {
  small = b;
  large = a;
}

for(i = 0; i < small.length; i++) {
  hash[small[i]] = 1;
}

for(i = 0; i < large.length; i++) {
  var l = large[i];
  if(hash[l]) {
    out.push(l)
  }
}

return out;
```

Reply:
... it might be nice to look at the lodash's implementation.